python_modules/dagster/dagster/_utils/cached_method.py

```diff
@@ -62,14 +63,26 @@ def a_method(self, arg1, arg2):
     """
     cache_attr_name = method.__name__ + CACHED_METHOD_FIELD_SUFFIX
 
+    # Cache these once self is first observed to avoid expensive work on each access
     arg_names = None
+    is_dagster_model = None
 
     @wraps(method)
     def _cached_method_wrapper(self: S, *args: P.args, **kwargs: P.kwargs) -> T:
-        if not hasattr(self, cache_attr_name):
-            cache: Dict[Hashable, T] = {}
-            setattr(self, cache_attr_name, cache)
+        nonlocal is_dagster_model
+        is_dagster_model = (
+            is_dagster_model if is_dagster_model is not None else isinstance(self, DagsterModel)
+        )
+
+        if is_dagster_model:
+            if cache_attr_name not in self._cached_method_cache:  # type: ignore
+                self._cached_method_cache[cache_attr_name] = {}  # type: ignore
+
+            cache = self._cached_method_cache[cache_attr_name]  # type: ignore
         else:
+            if not hasattr(self, cache_attr_name):
+                setattr(self, cache_attr_name, {})
+
```

Comment:
Are we retaining old behavior here in case someone is relying previous monkeypatching behavior? It might be worth the risk to just canonicalize on the `_cached_method_cache` property for all classes, not just `DagsterModel`?

Reply:
this cleaned things up quite a bit, good call